design/one-pager-change-logs.md

```diff
@@ -0,0 +1,310 @@
+that they manage, but this design will focus specifically on the observability
+of changes that the control plane makes to external systems at runtime. This is
+not intended to facilitate review or approve the changes before they are made,
+but rather to provide an auditable record of all changes that have been
+performed.
+
+This means that Crossplane (and its Providers) should start to expose data about
+all changes that the control plane is making to external systems, with as much
+detail as possible so the platform team can understand what is changing, for
+what reason it is changing, and the outcome of the change. With this
+information, Crossplane users will be able to build deeper trust and confidence
+in the operation of their control planes, as well as have forensic data
+available to them in the event that an unexpected change occurs.
+
+## Proposal
+
+### Change Log Entry Format
+
+The change log will be a series of individual entries that capture state about
+the resources involved in every operation a provider performs on its
+external system. Later in this document we will explore HOW the logs are
+generated and stored, but for now we will focus on WHAT will be collected.
+
+As a general principal, we will be collecting raw observation data that has not
+been processed in any significant way. A system of the users choosing that
+collects the logs entries can perform opinionated calculations, analysis, or
+alerting that is specific to their operating environment.
+
+Each entry in the change log will contain the following data:
+
+| Field Description | Example |
+|-------------------| ------- |
+| Timestamp | ISO 8601 format, e.g. `2023-04-01T12:34:56Z` |
+| Provider name | `xpkg.upbound.io/upbound/provider-aws-ec2:v1.8.0` |
+| Managed Resource identifying data (GVK/name) | `{apiVersion: ec2.aws.upbound.io/v1beta2, kind: Instance, name: dev-instance-bt66d}` |
+| Operation Type | `create\|update\|delete` |
+| Before operation desired/observed state of resource | `{full JSON dump of resource}` which includes desired `spec.forProvider` and observed `status.AtProvider` |
+| After operation desired/observed state of resource | `{full JSON dump of resource}` which includes desired `spec.forProvider` and observed `status.AtProvider`  |
+| Result of operation | `success` or error object |
+| (optional) Additional information that Providers can set as they choose | `{JSON object of arbitrary data}` |
+
+Note that we are capturing the full state of the resource both **before** and
+**after** the operation is performed. We could instead just record the "before"
+state during each reconciliation, but that would paint a less complete picture
+compared to observing the external system and recording the "after" state
+immediately after the operation as well. If we find during implementation and
+testing that recording the "after" state places too much burden on the system,
+we can revisit this decision.
```

Comment:
Some data point: we once did a benchmark on OpenShift with API auditing on with essentially the same data as you describe and measured 10% degradation.

Reply:
yes, upstream kube audit logs with full body of requests and responses.